library/ssl_msg.c

```diff
@@ -1163,45 +1134,13 @@ int mbedtls_ssl_decrypt_buf( mbedtls_ssl_context const *ssl,
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
     {
         padlen = 0;
-#if defined(MBEDTLS_USE_PSA_CRYPTO)
-        psa_status_t status;
-        size_t part_len;
-        psa_cipher_operation_t cipher_op = PSA_CIPHER_OPERATION_INIT;
-#endif /* MBEDTLS_USE_PSA_CRYPTO */
 
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
-        /* Skip psa decryption for null cipher */
-        if ( transform->psa_alg != MBEDTLS_SSL_NULL_CIPHER )
-        {
-            status = psa_cipher_decrypt_setup( &cipher_op,
-                                        transform->psa_key_dec, transform->psa_alg );
-
-            if( status != PSA_SUCCESS )
-                return( psa_ssl_status_to_mbedtls( status ) );
-
-            status = psa_cipher_set_iv( &cipher_op, transform->iv_dec, transform->ivlen );
-
-            if( status != PSA_SUCCESS )
-                return( psa_ssl_status_to_mbedtls( status ) );
-
-            status = psa_cipher_update( &cipher_op,
-                                        data, rec->data_len,
-                                        data, rec->data_len, &olen );
-
-            if( status != PSA_SUCCESS )
-                return( psa_ssl_status_to_mbedtls( status ) );
-
-            status = psa_cipher_finish( &cipher_op,
-                                        data + olen, rec->data_len - olen,
-                                        &part_len );
-
-            if( status != PSA_SUCCESS )
-                return( psa_ssl_status_to_mbedtls( status ) );
+    /* The only stream "cipher" we support is "NULL" */
+    if ( transform->psa_alg != MBEDTLS_SSL_NULL_CIPHER )
```

Comment:
Ok, looking at it again, and taking a step back:

1. We reach this point only if the test `if ( transform->psa_alg == MBEDTLS_SSL_NULL_CIPHER )` a few lines above succeeded, so testing again seems redundant - though we could move the comment to the test above.
2. In the non-PSA case, it's also true that the only stream "cipher" we support is "NULL", so we could skip the call to `mbedtls_cipher_crypt()` and merge the PSA and non-PSA cases.
3. Once we've done (2), the ` if( rec->data_len != olen )` check below becomes fully redundant and can be removed. (I'm not even sure if the `olen` variable is still needed.)

Reply:
I thought that we keep the null/stream cipher code because it is supported for non-psa build. What do you mean by "skip the call to mbedtls_cipher_crypt()"?
If I understand correctly your proposition, the we would end up with empty block:

```
#if defined(MBEDTLS_SSL_SOME_SUITES_USE_STREAM)
 /* The only stream "cipher" we support is "NULL" */
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if ( transform->psa_alg == MBEDTLS_SSL_NULL_CIPHER )
#else
    if( mode == MBEDTLS_MODE_STREAM )
#endif
    {
        MBEDTLS_SSL_DEBUG_MSG( 3, ( "before encrypt: msglen = %" MBEDTLS_PRINTF_SIZET ", "
                                    "including %d bytes of padding",
                                    rec->data_len, 0 ) );
    }
    else
#endif /* MBEDTLS_SSL_SOME_SUITES_USE_STREAM */
```

Reply:
What I'm saying is that the "null/stream cipher" code, since Mbed TLS 3.0 (and the removal of RC4 ciphersuites) is now "the null code" period, there is no longer any actual stream cipher that we support in TLS. This is regardless of whether PSA is used or not. So, what I'm saying here is that in the non-PSA case, calling `mbedtls_cipher_crypt()` when we know for a fact that the algorithm is `MBEDTLS_CIPHER_NULL` (because that's the only way `mode == MBEDTLS_MODE_STREAM` is true) is just a verbose and expensive way of performing the assignment `olen = rec->data_len`. (That was point 2 above.)

My other points are that further simplifications can be done then, in particular the check ` if( rec->data_len != olen )` is useless when it's directly preceded by `olen = rec->data_len` unconditonally; in fact the `olen` variable itself is now useless. So, overall, I think the whole block guarded by `MBEDTLS_SSL_SOME_SUITES_USE_STREAM` can be reduced to something like:
```c
#if defined(MBEDTLS_SSL_SOME_SUITES_USE_STREAM)
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if ( transform->psa_alg == MBEDTLS_SSL_NULL_CIPHER )
#else
    if( mode == MBEDTLS_MODE_STREAM )
#endif
    {
        MBEDTLS_SSL_DEBUG_MSG( 3, ( "before encrypt: msglen = %" MBEDTLS_PRINTF_SIZET ", "
                                    "including %d bytes of padding",
                                    rec->data_len, 0 ) );
        /* The only supported stream cipher is "NULL",
         * so there's nothing to do here.*/
    }
    else
#endif /* MBEDTLS_SSL_SOME_SUITES_USE_STREAM */
```
(Untested.) Also, similar simplifications apply to the decrypt function of course.

I know this PR is supposed to be about PSA, and the simplification of the non-PSA path could actually have been done before, but we didn't notice before, and I think it makes sense to simplify in this PR because it makes it easier to compare PSA to non-PSA code (otherwise we're left wondering why there's a call to `crypt()` function in one case but not the other).